org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/QualifiedNameReference.java

```diff
@@ -776,7 +776,7 @@ public TypeBinding getOtherFieldBindings(BlockScope scope) {
 				fieldReceiverType = fieldReceiverType.getErasureCompatibleType(field.declaringClass);// handle indirect inheritance thru variable secondary bound
 				FieldBinding originalBinding = previousField.original();
 				if (TypeBinding.notEquals(fieldReceiverType, oldReceiverType) || originalBinding.type.leafComponentType().isTypeVariable()) { // record need for explicit cast at codegen
-			    	setGenericCast(index-1,originalBinding.type.genericCast(fieldReceiverType)); // type cannot be base-type even in boxing case
+			    	setGenericCast(place,originalBinding.type.genericCast(fieldReceiverType)); // type cannot be base-type even in boxing case
```

Comment:
Wow this line stood unaffected by 18 years of maintenance!

As I'm not overly familiar with codegen, let me double check my newly acquired understanding by saying:

QNR uses two different kinds of indexes:
* array `tokens` refers to the syntactical structure which may include a *static prefix* (package, (static) classes)
* fields `other*` use different indexing because those arrays have no representation for the static prefix, but only relate to subsequent fields.

In the given example, we write into `otherGenericCasts` and hence we need the second kind of index, whereas `index` belongs to the former category.

@srikanth-sankaran  are we on the same page?

Reply:
> Wow this line stood unaffected by 18 years of maintenance!
> 
> As I'm not overly familiar with codegen, let me double check my newly acquired understanding by saying:
> 
> QNR uses two different kinds of indexes:
> 
> * array `tokens` refers to the syntactical structure which may include a _static prefix_ (package, (static) classes)
> * fields `other*` use different indexing because those arrays have no representation for the static prefix, but only relate to subsequent fields.
> 
> In the given example, we write into `otherGenericCasts` and hence we need the second kind of index, whereas `index` belongs to the former category.
> 
> @srikanth-sankaran are we on the same page?

Yes, that is a good summation.

Reply:
Perhaps it's worth a code comment to explain the different array lengths?

Anyway, not a reason to block the fix.